src/templates/partials/lighttpd.hbs

```diff
@@ -38,18 +19,40 @@ ssl.privkey = "/path/to/private_key"
 ssl.pemfile = "/path/to/signed_cert_followed_by_intermediates"
  {{#if (minver "1.0.2" form.opensslVersion)}}
   {{#if (minver "1.1.0" form.opensslVersion)}}
-ssl.openssl.ssl-conf-cmd = ("MinProtocol" => {{#if (includes "TLSv1" output.protocols)}}"TLSv1"{{else if (includes "TLSv1.1" output.protocols)}}"TLSv1.1"{{else if (includes "TLSv1.2" output.protocols)}}"TLSv1.2"{{else}}"TLSv1.3"{{/if}})
+   {{#if (includes "TLSv1" output.protocols)}}
+ssl.openssl.ssl-conf-cmd = ("MinProtocol" => "TLSv1")
+   {{else if (includes "TLSv1.1" output.protocols)}}
+ssl.openssl.ssl-conf-cmd = ("MinProtocol" => "TLSv1.1")
+   {{else if (includes "TLSv1.2" output.protocols)}}
+    {{#unless (minver "1.4.56" form.serverVersion)}}
+ssl.openssl.ssl-conf-cmd = ("MinProtocol" => "TLSv1.2")
+    {{/unless}}
+   {{else}}
+ssl.openssl.ssl-conf-cmd = ("MinProtocol" => "TLSv1.3")
+   {{/if}}
   {{else}}
 ssl.openssl.ssl-conf-cmd = ("Protocol" => "ALL, -SSLv2, -SSLv3{{#unless (includes "TLSv1" output.protocols)}}, -TLSv1{{/unless}}{{#unless (includes "TLSv1.1" output.protocols)}}, -TLSv1.1{{/unless}}{{#unless (includes "TLSv1.2" output.protocols)}}, -TLSv1.2{{/unless}}")
   {{/if}}
+  {{#if (minver "1.4.68" form.serverVersion)}}
+   {{#if output.serverPreferredOrder}}
+ssl.openssl.ssl-conf-cmd += ("Options" => "+ServerPreference")
+   {{/if}}
+  {{else}}
 ssl.openssl.ssl-conf-cmd += ("Options" => "{{#if output.serverPreferredOrder}}+{{else}}-{{/if}}ServerPreference")
+  {{/if}}
   {{#if output.ciphers.length}}
+   {{#if (minver "1.4.68" form.serverVersion)}}
+    {{#if (includes "old" form.config)}}
+ssl.openssl.ssl-conf-cmd += ("CipherString" => "{{{join output.ciphers ":"}}}")
+    {{/if}}
+   {{else}}
 # TLS modules besides mod_openssl might name ciphers differently
 # See https://redmine.lighttpd.net/projects/lighttpd/wiki/Docs_SSL
 ssl.openssl.ssl-conf-cmd += ("CipherString" => "{{{join output.ciphers ":"}}}")
+   {{/if}}
```

Comment:
This means not setting the cipher suites since ≥v1.4.68 unless `old`, i. e. `modern` and `intermediate` won't set the suites at all according to json specs? Isn't the intention of the configs to explicitly set the cipher suites according to the specs, instead of leaving it to the server defaults?

Reply:
@gene1wood or do I read it wrong? I understand there are more and more great defaults provided by the servers, but that doesn't mean the generator should give up and not output the cipher suites according to the current .json recommendations (which is whole another discussion TBH).

Is there a consensus when the actual output should be left out for server defaults behind a version conditional?

Reply:
@janbrasna as above, yes, it is better to use lighttpd secure TLS defaults.

Reply:
@janbrasna if you have further questions how to configure lighttpd securely, please refer to the [lighttpd TLS docs](https://wiki.lighttpd.net/Docs_SSL)

Reply:
@gstrauss Nope, I don't have further questions how to configure lighttpd securely, I don't think I need to. Surprisingly that's not the issue.

My question is about this change, that basically disallows the output of `https://statics.tls.security.mozilla.org/server-side-tls-conf-*.json` and leaves that to the vendor's definition. While claiming it's the @mozilla/server-side-tls's e. g. `intermediate` … when it is no longer so;)
_(even if the effectively used suites match today, that might not be true tomorrow — and that's why Mozilla's TLS specs are versioned and explicit, to be predictable, if needed for any given usecase e. g. PCI DSS testing etc.)_

This PR basically reads for `intermediate` and `latestVersion` i. e. the default in the UI: "Claim it's Mozilla's https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28recommended.29 while in reality supplying Lighty's https://redmine.lighttpd.net/projects/lighttpd/wiki/Docs_SSL#Perfect-Forward-Secrecy-PFS …"

There are more vendors wishing the output not to mess with their secure TLS defaults so I'll open a broader discussion for it. I understand it, servers with great defaults should get their credit for having safe defaults, yet I don't see the need to cripple the json specs output to achieve this.

Reply:
Quickly comparing the lists, lighty's includes CCM/8s & PSKs for embeds that are not too important here, but misses DHE-RSA-GCMs from server-side-tls-conf-* specs.

So the end output claiming to be @mozilla intermediate is in reality not that:

![Screen Shot 2024-01-05 at 23 11 31](https://github.com/mozilla/ssl-config-generator/assets/1784648/d5b21a28-3d5b-414f-a27c-259229eca93a)

See I'm no @april to have a say in this, I'm just testing the rendered output — and with no `configurations.intermediate.ciphers[]` present in the output at all, in the current state of affairs, that seemed like a bug. That's why I raised it for others to decide.

The need to somehow endorse great defaults where applicable is a whole another topic, unrelated to keeping the @mozilla's intermediate config output for lighty still to @mozilla's specs;)

Reply:
@janbrasna as you mentioned already, the mozilla TLS recommendation specification is probably due to for review, and I agree.

The last substantial updates to the ssl-config-generator specs were in 2020, when april published versions 5.3, 5.4, 5.5, and 5.6 (July 2020), **now 3 1/2 years ago**.  Gene published 5.7 in May 2023, with the simple change to add DHE-RSA-CHACHA20-POLY1305.

I am not suggesting whether or not dhe-rsa-aes128-gcm-sha256 or dhe-rsa-aes256-gcm-sha384 should be included.  I am saying that the spec should be reviewed to see if they should, or should not be removed from the intermediate cipher list.  Once that is refreshed (and not 3 1/2 years old), then we can have a more relevant conversation about whether or not ssl-config-generator should emit a cipher list to include those in the ssl-config-generator "Intermediate" lighttpd config.

One data point to compare with others: the current Cloudflare recommendations include only ciphers using elliptic curve diffie hellman ephemeral (ECDHE) for key exchange in both their "Modern" and "Compatible" cipher lists.  https://developers.cloudflare.com/ssl/reference/cipher-suites/recommendations/

Another data point is that I have not received any complaints about lighttpd TLS defaults since lighttpd updated its TLS defaults in Jan 2023 (after having announced the scheduled change throughout all of 2022).  I have received zero requests for lighttpd to include those dhe ciphers in the lighttpd TLS defaults, and I do not have any plans to add them.

Reply:
https://github.com/mozilla/server-side-tls/issues/285 notes "Note that DHE suites in TLS 1.2 are problematic." and discusses why Gene added DHE-RSA-CHACHA20-POLY1305 in ssl-config-generator 5.7.json

Reply:
Yes, for grading/testing consistency. Which is my point exactly with almost any similar change. If we promise a config tier, and a tool down the road uses that same definition to probe such config, and we fall short of the promise, that sucks.

E.g.. from another thread:

> "I humbly request that we prioritize updating this site to be at least as strict as the specifications, but stricter is OK."

Is presence of any CCM8 "stricter", or "at least as strict"? Who decides that? (RHEL folks would say "no", FWIW…)

So to avoid similar discussions, unless the usage of the tool warrants such one-off deviation from the recommendation data (think known issues with interop, key size or rotation etc.), I'd prefer not to challenge that HERE, but be helpful upstream, in defining what should the recommended profiles be, and align with other security folks from other domains adding another views to the issue as well (it's not all http servers and web browsers…). More secure defaults are only better for everyone, and pointing that out via comments is a good way to show no fine-tuning is really needed — but at the same time, however you may dislike the Mozilla TLS recommendations, this is only to turn them consistently into a config example, as little opinionated as possible TBH.

If you want to add a band-aid over the recommendations you don't agree with as a server author, it should be clearly stated though (for visibility, and to find a way folks are aware of).

Reply:
> however you may dislike the Mozilla TLS recommendations

That *is not* the issue.  The Mozilla TLS recommendations are outdated and currently appear to be unmaintained.  They have been supplanted by more up-to-date and maintained security recommendations from other qualified professional organizations.  One example: Cloudflare CDN https://developers.cloudflare.com/ssl/edge-certificates/additional-options/cipher-suites/recommendations/

As such, the Mozilla TLS recommendations, currently over 4 years old, should be reviewed and updated by Mozilla.  Until then, they should be discarded and ignored in favor of current and maintained TLS recommendations from other qualified professional organizations.  This is my professional opinion as a systems security programmer.  I would defend this statement to any CISO, CIO, or CTO if any project manager tried to pull "but look at *this* spec!"

Reply:
@gstrauss You're happy to ignore them as you wish, and also to work on any other organization's docs and tools (or your forks thereof) in the meantime… and I'll be happy to continue reviewing this changeset once the upstream config data match your expectations so you don't have to hardcode your disagreement in lieu of having to vandalize the outputs. (I'm perfectly fine with noting that in the output tho with the matching config included for reference for those envs that may need an exact match, so I'll look into that once I have some time to test drive a build with some of the recent changes…)

Reply:
But the more important bit is — you'd probably need to do some of your own testing to make sure at least the publicly available test suites, when comparing to "Mozilla intermediate" or similar profile for OK/NOK or grading, won't complain about CCM8 and PSK suites that you include extra.

(I don't know honestly. They are all sorta specific use, not sure I'd be recommending them for general use without any more explanation — and not sure if any e.g. PCI testing suite wouldn't complain, for reasons unknown to me TBH, such as the plaintext username transfer in "pure" PSK for ≤TLS1.2 perhaps?)

@gstrauss How do you expose these? Are they enumerated to be explicitly enabled, or do you use some addition/subtraction based on defaults, so if they were not in default strings on the OS level, you're not adding them anyway…? Or…?

Reply:
Read: If you need those, you need to know what to do and also enable them on purpose. We don't cater for such groups here in the configs. I'm not sure what all needs to be configured for them to even be able to end up negotiated — I'm only concerned about any automation/testing results that may stem from diverging from the exact restricted set in the jsons (as can be observed in the https://github.com/mozilla/server-side-tls/issues/285 rationale — following any allowlists perhaps too literally…)

Reply:
The omission of DHE-RSA-AES* is another topic, don't remember if just the DHE-RSA-CHACHA20 you provide is enough to support the sad SChannel compatibility when we had to keep the DHEs back in the day for RSA-only compatibility… will look into those, but quickly glancing https://learn.microsoft.com/en-us/windows/win32/secauthn/tls-cipher-suites-in-windows-7 it seems that would not support the oldest clients we claim unless ECDSA is used 🤷

Reply:
_(and this is the exact reason I'm not willing to burn too much time making the configs "too implicit" and having to hunt for compatibility woes one by one… when we have a set that makes the conservative compatibility match the guidelines exactly, unless noted otherwise in the comments)_

Reply:
I'm not really able to follow all the changes to Lighty defaults over time and with versions (hence why I'm hesitant to leave things to defaults and monitor changes constantly) but it seems that since the above comparison was posted, it's a bit stricter now — @gstrauss do I read it right that the PSK and DHE-RSA-CHACHA20 is gone now? And the presence of CCM/CCM8 is based on whether it was available in defaults before configuring the context? That would make the 3/24+ defaults reasonable, and only a note mentioning the absence of DHE thus the need for ECDSA chain to support the actual profile would suffice.

(BTW the actual Lighty cipher defaults are being documented under #PFS heading, it might be worth linking that section directly from the output comment as that's a valuable reference somewhat hard to find in general — as to why it's actually commented out, for comparison…)

**The main benefit of not configuring the cipherstring explicitly would be enabling compatibility with other TLS libraries — and that's a very important improvement!**